sdks/java/harness/src/test/java/org/apache/beam/fn/harness/data/QueueingBeamFnDataClientTest.java

```diff
@@ -0,0 +1,348 @@
+        Endpoints.ApiServiceDescriptor.newBuilder()
+            .setUrl(this.getClass().getName() + "-" + UUID.randomUUID().toString())
+            .build();
+    Server server =
+        InProcessServerBuilder.forName(apiServiceDescriptor.getUrl())
+            .addService(
+                new BeamFnDataGrpc.BeamFnDataImplBase() {
+                  @Override
+                  public StreamObserver<BeamFnApi.Elements> data(
+                      StreamObserver<BeamFnApi.Elements> outboundObserver) {
+                    outboundServerObserver.set(outboundObserver);
+                    waitForClientToConnect.countDown();
+                    return inboundServerObserver;
+                  }
+                })
+            .build();
+    server.start();
+    try {
+      ManagedChannel channel =
+          InProcessChannelBuilder.forName(apiServiceDescriptor.getUrl()).build();
+
+      BeamFnDataGrpcClient clientFactory =
+          new BeamFnDataGrpcClient(
+              PipelineOptionsFactory.create(),
+              (Endpoints.ApiServiceDescriptor descriptor) -> channel,
+              OutboundObserverFactory.trivial());
+      QueueingBeamFnDataClient queueingClient = new QueueingBeamFnDataClient(clientFactory);
+
+      InboundDataClient readFutureA =
+          queueingClient.receive(
+              apiServiceDescriptor,
+              ENDPOINT_A,
+              CODER,
+              (WindowedValue<String> wv) -> {
+                throw new InterruptedException("Intentionally Fail"); // Error injected here.
+              });
+
+      waitForClientToConnect.await();
+
+      Future<?> sendElementsFuture =
+          executor.submit(
+              () -> {
+                outboundServerObserver.get().onNext(ELEMENTS_A_1);
+                // Purposefully transmit some data before the consumer for B is bound showing that
+                // data is not lost
+                outboundServerObserver.get().onNext(ELEMENTS_B_1);
+              });
+
+      InboundDataClient readFutureB =
+          queueingClient.receive(
+              apiServiceDescriptor,
+              ENDPOINT_B,
+              CODER,
+              (WindowedValue<String> wv) -> {
+                inboundValuesB.add(wv);
+              });
+
+      Future<?> drainElementsFuture =
+          executor.submit(
+              () -> {
+                boolean intentionallyFailed = false;
+                try {
+                  queueingClient.drainAndBlock();
+                } catch (InterruptedException e) {
+                  intentionallyFailed = true;
+                } catch (Exception e) {
+                  LOG.error("Unintentional failure", e);
+                  fail();
+                }
+                assertTrue(intentionallyFailed);
+              });
+
+      boolean intentionallyFailedA = false;
+      try {
+        readFutureA.awaitCompletion(); // All InboundDataClients should be completed.
+      } catch (ExecutionException e) {
```

Comment:
You can check that the execution exception has a cause which is your runtime exception, ditto here and below.